### PR TITLE
john: Fix installation process, add test

### LINKED ISSUE #47164: john: fopen: /usr/local/share/john/john.conf: No such file or directory
# Please note we will close your issue without comment if you delete, do not read or do not fill out the issue checklist below and provide ALL the requested information. If you repeatedly fail to use the issue template, we will block you from ever submitting issues to Homebrew again.

- [x] ran `brew update` and can still reproduce the problem?
- [x] ran `brew doctor`, fixed all issues and can still reproduce the problem?
- [x] ran `brew gist-logs <formula>` (where `<formula>` is the name of the formula that failed) and included the output link?
- [x] if `brew gist-logs` didn't work: ran `brew config` and `brew doctor` and included their output with your issue?

## What you were trying to do (and why)

Install `john` formula, and run it

## What happened (include command output)

I run `john` command, and see this:

<details>
  <summary>Command output</summary>
  <pre>
$ john
fopen: /usr/local/share/john/john.conf: No such file or directory
  </pre>
</details>

## What you expected to happen

No error, and john runs like this:

```
$ john
John the Ripper password cracker, version 1.9.0
Copyright (c) 1996-2019 by Solar Designer
```

## Step-by-step reproduction instructions (by running `brew install` commands)

Just run:

```shell
brew install john
```
It download a bottle and installed as normal. And then:

```
$ john
fopen: /usr/local/share/john/john.conf: No such file or directory
```

As i checked it, `/usr/local/share/john` not even exist. And this is how i fix it manually:

```shell
cd /usr/local/share
ln -s /usr/local/Cellar/john/1.9.0/libexec john
cd john
cp john.ini john.conf
```

and now it seem works.

My versions:

```
$ brew config
HOMEBREW_VERSION: 2.1.16
ORIGIN: https://github.com/Homebrew/brew
HEAD: 3aa7624284c43180a3f3a71aeaa9263092868e12
Last commit: 3 weeks ago
Core tap ORIGIN: https://github.com/Homebrew/homebrew-core
Core tap HEAD: 992e817c65f5ab22df5d7f1eb08e38924eeafd9a
Core tap last commit: 62 minutes ago
HOMEBREW_PREFIX: /usr/local
HOMEBREW_NO_BOTTLE_SOURCE_FALLBACK: 1
CPU: dodeca-core 64-bit kabylake
Homebrew Ruby: 2.6.3 => /usr/local/Homebrew/Library/Homebrew/vendor/portable-ruby/2.6.3/bin/ruby
Clang: 10.0 build 1001
Git: 2.24.0 => /usr/local/bin/git
Curl: 7.54.0 => /usr/bin/curl
Java: 1.8.0_202
macOS: 10.14.6-x86_64
CLT: 10.3.0.0.1.1562985497
Xcode: N/A
CLT headers: 10.3.0.0.1.1562985497
```

Discussion before at discourse.brew.sh: [https://discourse.brew.sh/t/john-fopen-usr-local-share-john-john-conf-no-such-file-or-directory-is-it-reproducible/6328](https://discourse.brew.sh/t/john-fopen-usr-local-share-john-john-conf-no-such-file-or-directory-is-it-reproducible/6328)

### PR DESCRIPTION
**Pull request by gromgit**
* Converted config-by-patch to `inreplace` block
* Removed config file renaming that actually broke things
* Added HOME directory patch that was integrated into official `john-jumbo` fork, but missing from core
* Installed main and symlinked binaries to `bin`, rather than symlinking
* Added test block

Fixes #47164

- [x] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/HEAD/CONTRIBUTING.md)?
- [x] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [x] Have you built your formula locally with `brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Is your test running fine `brew test <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Does your build pass `brew audit --strict <formula>` (after doing `brew install <formula>`)?

-----

### CODE REVIEW

File: Formula/john.rb [hunk 1]
```diff
@@ -14,45 +16,85 @@ class John < Formula
 
   conflicts_with "john-jumbo", :because => "both install the same binaries"
 
-  patch :DATA # Taken from MacPorts, tells john where to find runtime files
+  patch :DATA # Backporting official patch from jumbo fork (https://www.openwall.com/lists/john-users/2016/01/04/1)
```
Comment (issyl0) [inline Formula/john.rb:19]: As this patch is quite long, could we put it in [Homebrew/formula-patches](https://github.com/Homebrew/formula-patches) and pull it in with a `patch do` block?

**Review by issyl0 — COMMENTED** [review]
Thanks for having a go at fixing this - and we greatly appreciate the test! 🎉 Just one tiny comment on the location of the patch...

File: Formula/john.rb (hunk 1 above)
Comment (gromgit) [reply Formula/john.rb:19]: Sure, but how do I go about doing this? The README there and the main docs say nothing about the required workflow to do so.

Comment (SMillerDev) [reply Formula/john.rb:19]: Make a pull request to add it. That's all really

Comment (issyl0) [reply Formula/john.rb:19]: https://github.com/Homebrew/formula-patches/pull/286 is merged now, so all that's left is to add `patch do` with the URL to the patch and its SHA256. There's [plenty of prior art](https://github.com/Homebrew/homebrew-core/search?q=Homebrew%2Fformula-patches&unscoped_q=Homebrew%2Fformula-patches).

Comment (gromgit) [reply Formula/john.rb]: OK, done.

**Review by fxcoudert — APPROVED** [review]

**Pull request closed without merging**